src/main/webapp/js/projects/react/components/ExpandedProject.js

```diff
@@ -12,12 +14,50 @@ export default function ExpandedProject(props) {
     );
   }
 
-  // TODO : Add expanded project frontend view.
-  // For now this is just the name of the project to ensure that everything
-  // works.
+  const project = props.project;
+  const projectTags = project.previewData.topicTags;
+  // Adds the primary language to the list of project tags. Should add
+  // different types of tags in the future to make it easier to differentiate
+  // topics from languages.
+  projectTags.unshift(project.previewData.primaryLanguage);
+
   return (
-    <div>
-      {props.project.previewData.name}
+    <div className="container-fluid">
+      <div></div>
+      <div className="row mt-4">
+        <div className="col-md-9">
+          <div className="card mx-1 mb-2">
+            <div className="card-body px-3 py-2">
+              <h1 className="card-title emphasis">
+                {project.previewData.name}
+              </h1>
+              <TagList tags={projectTags} />
+            </div>
+          </div>
+          <div className="ml-1 emphasis">Mentors:</div>
+          <div className="d-flex flex-wrap mb-2">
+            {/* TODO : Use the mentor User ID as the key. Once it is
+                available. */}
```

Comment:
Just wondering, what are the benefits of using the mentor user ID as key? IIUC key in react is just for rendering something correctly and an index should be just fine. I suppose we could remove 1 very tightly scoped variable?

Reply:
Good question! Using the index in the list as the ID works, and it is what React defaults to if you don't manually set keys, but it actually is considered [bad practice](https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318). Here is a snippet from [React's official documentation](https://reactjs.org/docs/reconciliation.html#recursing-on-children):

> When children have keys, React uses the key to match children in the original tree with children in the subsequent tree.

React uses keys to keep track of component children such that it only needs to re-render the items that have changed. You can imagine that if the ordering changes (like for example the order is reversed) React assumes that nothing has changed because the keys (if they are indices) have stayed the same, and therefore nothing is re-rendered and you have a problem.

> The best way to pick a key is to use a string that uniquely identifies a list item among its siblings. Most often you would use IDs from your data as keys.

React's recommended way to solve this problem is to use some sort of ID that can uniquely identify the data being represented by the component, which is in this case the User's ID.

Reply:
Sounds good, thanks for the details and links!!